### PR TITLE
 Fixed #41 by using BeanSerializer as base for HibernateProxySerializ…

### LINKED ISSUE #41: Using Hibernate Module for lazy loading does not work with @JsonIdentityInfo
I'm not sure if this is because the proxy does not have the annotation on it or if it is something else.

We have several entities with cyclic dependencies and were originally defined as EAGER loading.  We added the @JsonIdentityInfo on those classes and they were serializing fine.

I changed several of them to LAZY loading for performance reasons, and now it seems that when the classes that are proxied are serialized, the @JsonIdentityInfo is not being honored by the Hibernate 4 Module.

We are using JBoss 7.2.0 with RestEasy.  I added the appropriate dependencies to my Maven pom, wrote a custom provider and simply instantiated an ObjectMapper and registered the Hibernate4Module.

When I retrieve the object via a call to RestEasy, it attempt to serialize it using the ObjectMapper I created, but it ends with an infinite loop and eventually throws an exception.  Similar to the way Jackson 1 used to react with cyclic dependencies.  It prints the same two error lines multiple times in the stack trace (I am cutting it off after a few....it's probably several hundred times repeated:

```
23:46:41,588 ERROR [org.apache.catalina.core.ContainerBase.[jboss.web].[default-host].[/mission_planning_war].[Resteasy]] (http-/127.0.0.1:8080-1) JBWEB000236: Servlet.service() for servlet Resteasy threw exception: java.lang.ClassNotFoundException: com.fasterxml.jackson.databind.JsonMappingException$Reference from [Module "deployment.elint_ear-0.0.1-SNAPSHOT.ear.mission_planning_war-0.0.1-SNAPSHOT.war:main" from Service Module Loader]
at org.jboss.modules.ModuleClassLoader.findClass(ModuleClassLoader.java:190) [jboss-modules.jar:1.2.0.CR1]
at org.jboss.modules.ConcurrentClassLoader.performLoadClassUnchecked(ConcurrentClassLoader.java:468) [jboss-modules.jar:1.2.0.CR1]
at org.jboss.modules.ConcurrentClassLoader.performLoadClassChecked(ConcurrentClassLoader.java:456) [jboss-modules.jar:1.2.0.CR1]
at org.jboss.modules.ConcurrentClassLoader.performLoadClass(ConcurrentClassLoader.java:398) [jboss-modules.jar:1.2.0.CR1]
at org.jboss.modules.ConcurrentClassLoader.loadClass(ConcurrentClassLoader.java:120) [jboss-modules.jar:1.2.0.CR1]
at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:613) [jackson-databind-2.2.1.jar:2.2.1]
at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:142) [jackson-databind-2.2.1.jar:2.2.1]
at com.fasterxml.jackson.datatype.hibernate4.HibernateProxySerializer.serialize(HibernateProxySerializer.java:75) [jackson-datatype-hibernate4-2.2.1.jar:2.2.1]
at com.fasterxml.jackson.datatype.hibernate4.HibernateProxySerializer.serialize(HibernateProxySerializer.java:24) [jackson-datatype-hibernate4-2.2.1.jar:2.2.1]
at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:569) [jackson-databind-2.2.1.jar:2.2.1]
at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:597) [jackson-databind-2.2.1.jar:2.2.1]
at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:142) [jackson-databind-2.2.1.jar:2.2.1]
at com.fasterxml.jackson.databind.ser.std.CollectionSerializer.serializeContents(CollectionSerializer.java:117) [jackson-databind-2.2.1.jar:2.2.1]
at com.fasterxml.jackson.databind.ser.std.CollectionSerializer.serializeContents(CollectionSerializer.java:23) [jackson-databind-2.2.1.jar:2.2.1]
at com.fasterxml.jackson.databind.ser.std.AsArraySerializerBase.serialize(AsArraySerializerBase.java:186) [jackson-databind-2.2.1.jar:2.2.1]
at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:569) [jackson-databind-2.2.1.jar:2.2.1]
at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:597) [jackson-databind-2.2.1.jar:2.2.1]
at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:142) [jackson-databind-2.2.1.jar:2.2.1]
at com.fasterxml.jackson.datatype.hibernate4.HibernateProxySerializer.serialize(HibernateProxySerializer.java:75) [jackson-datatype-hibernate4-2.2.1.jar:2.2.1]
at com.fasterxml.jackson.datatype.hibernate4.HibernateProxySerializer.serialize(HibernateProxySerializer.java:24) [jackson-datatype-hibernate4-2.2.1.jar:2.2.1]
at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:569) [jackson-databind-2.2.1.jar:2.2.1]
at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:597) [jackson-databind-2.2.1.jar:2.2.1]
at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:142) [jackson-databind-2.2.1.jar:2.2.1]
at com.fasterxml.jackson.databind.ser.std.CollectionSerializer.serializeContents(CollectionSerializer.java:117) [jackson-databind-2.2.1.jar:2.2.1]
at com.fasterxml.jackson.databind.ser.std.CollectionSerializer.serializeContents(CollectionSerializer.java:23) [jackson-databind-2.2.1.jar:2.2.1]
at com.fasterxml.jackson.databind.ser.std.AsArraySerializerBase.serialize(AsArraySerializerBase.java:186) [jackson-databind-2.2.1.jar:2.2.1]
at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:569) [jackson-databind-2.2.1.jar:2.2.1]
at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:597) [jackson-databind-2.2.1.jar:2.2.1]
```

### PR DESCRIPTION
**Pull request by ManuelB**
…er and using BeanSerializerFactory as base for HibernateSerializers

See Bug #41

### CODE REVIEW

**Comment by cowtowncoder** [comment]
Thank you for the contribution.

Is there any chance you could avoid reformatting the code wrt indentation? Project uses 4-space indentation, and changing it will make it quite difficult to read the actual changes. So the usual process is to just follow existing indentation and naming convention; changes in sections that are new are ok (we can reformat them if need be) as they do not affect readability wrt changes.

Now, as to changes: I do not think proxy serializer should extend `BeanSerializer` (or `BeanSerializerBase`). While I understand that may be convenient, it is not a good approach in my opinion, since proxies are not beans: they are referential types. In general, these classes are not meant to be extended by non-core code, meaning that no attempt is made to keep changes compatible for such use case. Developers are of course free to extend these classes if they choose to, knowing the risks. But I don't want modules to do that unless there is absolutely no other way to achieve certain behavior. I am not convinced it is necessary here, yet.

**Comment by ManuelB** [comment]
Hi,
actually I am directly using the indention feature of eclipse and I can't revert these easily to another format.

I just added some functions to exclude hibernate specific properties (handler, hibernateInitializer).

I fully agree that the way how it is implemented is not that nice but hibernate entities are magical objects that a modified by the class loader.

Unfortunately I currently don't have the necessary time budget to find a more elegant solution. I provided some test cases that are proofing that it works. Maybe if I have time during the next 6-12 month I might revisit the code maybe the JSON spec from Java will make it easier to serialize such objects.

Hope that helps
    Manuel

**Comment by reda-alaoui** [comment]
Hello,

First of all, thank you for your pull request.
I was using `2.5.1` and each time the object to serialize was an instance of HibernateProxy and initialised, **all jackson annotations were ignored**.

I merged this commit in my own fork.
Since then, I am facing with the following issue:

```
Caused by: com.fasterxml.jackson.databind.JsonMappingException: 
No serializer found for class org.hibernate.proxy.pojo.javassist.JavassistLazyInitializer 
and no properties discovered to create BeanSerializer 
(to avoid exception, disable SerializationFeature.FAIL_ON_EMPTY_BEANS)
```

Did you also faced this issue?

**Comment by reda-alaoui** [comment]
I disabled `SerializationFeature.FAIL_ON_EMPTY_BEANS`, everything seems to work well now.

**Comment by reda-alaoui** [comment]
Me again,

Since this commit, I am having the following error:
```
GRAVE: "Servlet.service()" pour la servlet main a généré une exception
com.fasterxml.jackson.databind.JsonMappingException: could not initialize proxy - no Session (through reference chain: com.foo.app.grid.GridConfiguration["user"]->com.foo.app.user.factory.User["siteClearances"]->org.hibernate.collection.internal.PersistentSet[0]->com.foo.app.site.SiteClearance["site"]->com.foo.app.site.Site["company"]->com.foo.app.company.Company_$$_jvstebb_7["name"])
	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:210)
	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:177)
	at com.fasterxml.jackson.databind.ser.std.StdSerializer.wrapAndThrow(StdSerializer.java:190)
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:671)
	at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:156)
	at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:575)
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:663)
	at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:156)
	at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:575)
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:663)
	at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:156)
	at com.fasterxml.jackson.databind.ser.std.CollectionSerializer.serializeContents(CollectionSerializer.java:132)
	at com.fasterxml.jackson.databind.ser.std.CollectionSerializer.serialize(CollectionSerializer.java:94)
	at com.fasterxml.jackson.databind.ser.std.CollectionSerializer.serialize(CollectionSerializer.java:24)
	at com.fasterxml.jackson.datatype.hibernate4.PersistentCollectionSerializer.serialize(PersistentCollectionSerializer.java:125)
	at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:575)
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:663)
	at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:156)
	at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:575)
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:663)
	at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:156)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:129)
	at com.fasterxml.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:2240)
	at org.springframework.web.servlet.view.json.MappingJackson2JsonView.writeContent(MappingJackson2JsonView.java:302)
	at org.springframework.web.servlet.view.json.MappingJackson2JsonView.renderMergedOutputModel(MappingJackson2JsonView.java:257)
	at org.springframework.web.servlet.view.AbstractView.render(AbstractView.java:264)
	at org.springframework.web.servlet.DispatcherServlet.render(DispatcherServlet.java:1201)
	at org.springframework.web.servlet.DispatcherServlet.processDispatchResult(DispatcherServlet.java:986)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:933)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:851)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:953)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:855)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:644)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:829)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:725)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:291)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:100)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:118)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:84)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:113)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:113)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:154)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:199)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
	at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:50)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:106)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:87)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:192)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:160)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:343)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:260)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
        ....
Caused by: org.hibernate.LazyInitializationException: could not initialize proxy - no Session
	at org.hibernate.proxy.AbstractLazyInitializer.initialize(AbstractLazyInitializer.java:165)
	at org.hibernate.proxy.AbstractLazyInitializer.getImplementation(AbstractLazyInitializer.java:286)
	at org.hibernate.proxy.pojo.javassist.JavassistLazyInitializer.invoke(JavassistLazyInitializer.java:185)
	at com.foo.app.company.Company_$$_jvstebb_7.getName(Company_$$_jvstebb_7.java)
	at sun.reflect.GeneratedMethodAccessor147.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:536)
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:663)
	... 100 more
```

**Comment by reda-alaoui** [comment]
In fact, your commit can't be merged back to 2.5.
Sorry for the false alarm!

**Comment by ManuelB** [comment]
Hi Reda,
sorry, but I can't support my experimental feature. 

It seems that you detached you entity from the persistence context before desirializing.

You might want to use JBoss 7 EAP which is a supported Java EE 6 product.

/Manuel

**Comment by reda-alaoui** [comment]
Hi Manuel,

If serialization must be performed in an active hibernate session, doesn't that mean that lazy loaded objects are systematically intialized by the new HibernateProxySerializer?

LazyLoadingTest` only checks that persistent collections are not initialized by serialization.
It doesn't check for `HibernateProxy` initialisation triggered by serialization.

**Comment by ManuelB** [comment]
@reda-alaoui yes, the new hibernate serializer will initialize all objets. If you do not want that you have to annotate them with @XmlIDREF:

https://docs.oracle.com/javaee/5/api/javax/xml/bind/annotation/XmlIDREF.html

File: hibernate4/src/test/java/com/fasterxml/jackson/datatype/hibernate4/HibernateTest.java [hunk 1]
```diff
@@ -90,6 +91,7 @@ public void testCyclesJson() throws Exception {
         
         // false -> no forcing of lazy loading
         ObjectMapper mapper = mapperWithModule(false);
+        mapper.disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
```
Comment (cowtowncoder) [inline hibernate4/src/test/java/com/fasterxml/jackson/datatype/hibernate4/HibernateTest.java]: But wouldn't this hide possible errors? It should not be necessary to do this.

**Comment by cowtowncoder** [comment]
I don't think this PR is active any more, closing.

**Pull request closed without merging**